src/aws-cpp-sdk-core/source/auth/GeneralHTTPCredentialsProvider.cpp

```diff
@@ -0,0 +1,204 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+
+#include <aws/core/auth/GeneralHTTPCredentialsProvider.h>
+
+#include <aws/core/config/AWSProfileConfigLoader.h>
+#include <aws/core/platform/Environment.h>
+#include <aws/core/utils/logging/LogMacros.h>
+#include <aws/core/utils/json/JsonSerializer.h>
+#include <fstream>
+
+using namespace Aws::Auth;
+
+static const int AWS_CREDENTIAL_PROVIDER_EXPIRATION_GRACE_PERIOD = 5 * 1000;
+static const char GEN_HTTP_LOG_TAG[] = "GeneralHTTPCredentialsProvider";
+
+const char GeneralHTTPCredentialsProvider::AWS_CONTAINER_AUTHORIZATION_TOKEN_FILE[] = "AWS_CONTAINER_AUTHORIZATION_TOKEN_FILE";
+const char GeneralHTTPCredentialsProvider::AWS_CONTAINER_CREDENTIALS_RELATIVE_URI[] = "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI";
+const char GeneralHTTPCredentialsProvider::AWS_CONTAINER_CREDENTIALS_FULL_URI[]     = "AWS_CONTAINER_CREDENTIALS_FULL_URI";
+const char GeneralHTTPCredentialsProvider::AWS_CONTAINER_AUTHORIZATION_TOKEN[]      = "AWS_CONTAINER_AUTHORIZATION_TOKEN";
+
+
+bool ShouldCreateGeneralHTTPProvider(const Aws::String& relativeUri, const Aws::String& absoluteUri, const Aws::String authToken)
+{
+    if (authToken.find("\r\n") != Aws::String::npos)
+    {
+        AWS_LOGSTREAM_WARN(GEN_HTTP_LOG_TAG, "Can't use General HTTP Provider: AWS_CONTAINER_AUTHORIZATION_TOKEN env value contains invalid characters (\\r\\n)");
+        return false;
+    }
+
+    if (!relativeUri.empty())
+    {
+        // The provider MAY choose to assert syntactical validity of the resulting URI
+        // perform very basic check here
+        if (relativeUri[0] != '/') {
+          AWS_LOGSTREAM_WARN(GEN_HTTP_LOG_TAG, "Can't use General HTTP Provider: AWS_CONTAINER_CREDENTIALS_RELATIVE_URI does not begin with /");
+          return false;
+        } else {
+          // full URI is not used in case of a relative one present
+          return true;
+        }
+    }
+
+    if (!absoluteUri.empty())
+    {
+        // If the resolved URI’s scheme is HTTPS, its hostname may be used in the request
+        if (Aws::Utils::StringUtils::ToLower(absoluteUri.c_str()).rfind(Aws::String("https://"), 0) == 0) // if starts_with
+        {
+            return true;
+        }
+
+        Aws::Http::URI absUri(absoluteUri);
+        const Aws::String& authority = absUri.GetAuthority();
+
+        // Otherwise, implementations MUST fail to resolve when the URI hostname does not satisfy any of the following conditions
+        // address is an ECS / EKS container host
+        if (authority == "169.254.170.2") // ECS container host
```

Comment:
nit: add constants for string that seem random
169.254.170.2 -> `ECS_CONTAINER_HOST`
169.254.170.23 -> `EKS_CONTAINER_HOST`
fd00:ec2::23 -> `EKS_IPV6`